framework/sandstone.cpp

```diff
@@ -2828,6 +2829,89 @@ static void warn_deprecated_opt(const char *opt)
             program_invocation_name, opt);
 }
 
+/* Setup of the performance counters we read to get getloadavg() on linux. */
+#ifdef _WIN32
+
+/* Performance counters we are going to read */
+static const char PQL_COUNTER_PATH[] = "\\System\\Processor Queue Length";
+static HCOUNTER pql_counter;
+
+static const char PT_COUNTER_PATH[]  = "\\Processor(_Total)\\% Processor Time";
+static HCOUNTER pt_counter;
+
+
+static constexpr unsigned TOTAL_5MIN_SAMPLES_COUNT = ((5u * 60u) / 5u);
+static constexpr unsigned SAMPLE_INTERVAL_SECONDS = 5u;
+
+static std::atomic<double> loadavg = 0;
+
+static void loadavg_windows_callback(PVOID, BOOLEAN)
+{
+    PDH_FMT_COUNTERVALUE vpql, vpt;
+
+    if (PdhGetFormattedCounterValue((PDH_HCOUNTER)pql_counter, PDH_FMT_DOUBLE, 0, &vpql) != ERROR_SUCCESS) {
+        return;
+    }
+
+    if (PdhGetFormattedCounterValue((PDH_HCOUNTER)pt_counter, PDH_FMT_DOUBLE, 0, &vpt) != ERROR_SUCCESS) {
+        return;
+    }
+
+    // We calculate current average load as average instantaenous cpu load plus amount of
+    // tasks that are ready to run but cannot be scheduled because CPUs are already
+    // running other tasks.
+    //
+    // We divide by 100.0 to get value in range (0.0;1.0) instead of percents.
+    //
+    // We also mutliply by number of cpus to make the metric behave more like the
+    // /proc/loadavg from Linux, so we get value from range (0.0;num_cpus()), where
+    // num_cpus() value means all cores at 100% utilization.
+    const double current_avg_cpu_usage = (vpt.doubleValue * num_cpus() / 100.0);
+    const double current_proc_queue    = vpql.doubleValue;
+    const double current_avg_load      = current_avg_cpu_usage + current_proc_queue;
+
+    double loadavg_ = loadavg.load(std::memory_order::memory_order_consume);
+    loadavg_ = (current_avg_load + ((TOTAL_5MIN_SAMPLES_COUNT - 1) * loadavg_)) / TOTAL_5MIN_SAMPLES_COUNT;
+
+    loadavg.store(loadavg_, std::memory_order::memory_order_release);
+}
+
+static int setup_windows_loadavg_perf_counters()
+{
+    HQUERY load_query;
+    HANDLE load_event;
+
+    if (PdhOpenQueryA(NULL, 0, &load_query) != ERROR_SUCCESS)
```

Comment:
We need the query to stay opened in the kernel, because we register callback with `PdhCollectQueryDataEx` that will be called when `load_event` fires associated with this query. It's closed by the kernel when the dcdiag exit.